sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,232 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An outbound data buffering aggregator with size-based buffer and time-based buffer if
+ * corresponding options are set.
+ *
+ * <p>The default size-based buffer threshold can be overridden by specifying the experiment {@code
+ * data_buffer_size_limit=<bytes>}
+ *
+ * <p>The default time-based buffer threshold can be overridden by specifying the experiment {@code
+ * data_buffer_time_limit_ms=<milliseconds>}
+ */
+@SuppressWarnings({
+  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+public class BeamFnDataOutboundAggregator implements AutoCloseable {
+
+  public static final String DATA_BUFFER_SIZE_LIMIT = "data_buffer_size_limit=";
+  public static final int DEFAULT_BUFFER_LIMIT_BYTES = 1_000_000;
+  public static final String DATA_BUFFER_TIME_LIMIT_MS = "data_buffer_time_limit_ms=";
+  public static final long DEFAULT_BUFFER_LIMIT_TIME_MS = -1L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(BeamFnDataOutboundAggregator.class);
+  private final int sizeLimit;
+  private final long timeLimit;
+  private final Map<LogicalEndpoint, Coder<?>> outputLocations;
+  private final StreamObserver<Elements> outboundObserver;
+  private final Map<LogicalEndpoint, ByteString.Output> buffer;
+  @VisibleForTesting ScheduledFuture<?> flushFuture;
+  private long byteCounter;
+
+  public BeamFnDataOutboundAggregator(
+      PipelineOptions options, StreamObserver<BeamFnApi.Elements> outboundObserver) {
+    this.sizeLimit = getSizeLimit(options);
+    this.timeLimit = getTimeLimit(options);
+    this.outputLocations = new HashMap<>();
+    this.buffer = new ConcurrentHashMap<>();
+    this.outboundObserver = outboundObserver;
+    if (timeLimit > 0) {
+      this.flushFuture =
+          Executors.newSingleThreadScheduledExecutor(
+                  new ThreadFactoryBuilder()
+                      .setDaemon(true)
+                      .setNameFormat("DataBufferOutboundFlusher-thread")
+                      .build())
+              .scheduleAtFixedRate(
+                  this::periodicFlush, timeLimit, timeLimit, TimeUnit.MILLISECONDS);
+    } else {
+      this.flushFuture = null;
+    }
+  }
+
+  /**
+   * Register the outbound logical endpoint alongside it's value coder. All registered endpoints
+   * will eventually be removed when {@link #close()} is called at the end of a bundle.
+   */
+  public <T> void registerOutputLocation(LogicalEndpoint endpoint, Coder<T> coder) {
+    LOG.debug("Registering endpoint: {}", endpoint);
+    outputLocations.put(endpoint, coder);
+  }
+
+  public void flush() throws IOException {
+    if (byteCounter > 0) {
+      outboundObserver.onNext(convertBufferForTransmission().build());
+    }
+  }
+
+  /**
+   * Closes the streams for all registered outbound endpoints. Should be called at the end of each
+   * bundle.
+   */
+  @Override
+  public void close() {
```

Comment:
I'm not sure you gain much by making this AutoCloseable since your next PR will be `steal output and do nothing if no data has yet to be sent` so it may make more sense to choose a clearer name here then `close()`.

Reply:
done.